### ISSUE TITLE
Completions should be coming from goflow

### ISSUE THREAD

**Issue by nicpottier** [rapidpro/rapidpro#974]
The set of possible context variables should be generated from some sort of JSON file that goflow provides. That lets us add things at the goflow side and have the editor do the right thing with respect to completion.

- [x] https://github.com/nyaruka/floweditor/issues/650
- [x] https://github.com/nyaruka/goflow/pull/721

**Comment by rowanseymour** [comment]
Ok so https://github.com/nyaruka/goflow/pull/721 is a bit experimental but this is what I've come up with so far. The help strings aren't final, just me trying to fill everything out..

* https://github.com/nyaruka/goflow/blob/context/docs/completion.json
* https://github.com/nyaruka/goflow/blob/context/docs/completion.txt (flatted view of all paths)

Some things to consider

* Not sure how array properties should work. Right now I don't think the editor's auto-completion has any concept of arrays.
* It's tricky but not impossible (via the idea of templatable help strings) to make the help strings context dependent. For example when completing `@parent.contact.na` it could be `the name of the contact`, `the name of a the parent run's contact`, `the name of the parent contact`. The latter could be generated by allowing the help text of `Contact.name` to be a template like `the name of the {parent} contact` where `{parent}` is the key before `contact`.
* When auto-completing `@conta` should it be the `the name of the contact` because that's what `@contact` would render as, or should it be `the contact` to make it clearer that it's not just a name and it has sub-properties.. I lean toward the latter.

**Comment by nicpottier** [comment]
Nice, I like this.

To your third point, aren't we always asking ourselves what these complex things resolve to? (IE, does `@results.foo` return the category name or the value?) To me that's an argument to be explicit (`the name of the contact`). That it is a more complex thing will fall out from the completion having child elements maybe? Argument against the above is that `@(json(contact))` does not give you `Fred` but instead something complex. But maybe we file that into advanced user problems?

**Comment by nicpottier** [comment]
In the json what does `ref` refer to in `type_ref`? Any reason that isn't just simply `type`?

**Comment by rowanseymour** [comment]
That's fair point re defaults.. tho it would be pretty weird for some other objects with defaults - for example should help for `@ru` be `the name of the contact and UUID of the flow`. Ideally maybe you'd see `@contac` as both `the current contact` and an option to auto-complete as `@contact` as `the name of the contact`. Not sure.

Re `type_ref`, I had that but it seemed a little confusing because actual types in the JSON have a `type` field to indicate that they are a declaration of that type. Properties on the otherhand aren't declarations of types. 

Maybe it would be clearer if the properties use `type`, the type declarations use `name` or `id`? Just trying to keep those things distinguishable.

**Comment by nicpottier** [comment]
Oh I see.. do they need to be different at all? Aren't `contact` and `text` both types fundamentally?

**Comment by nicpottier** [comment]
> That's fair point re defaults.. tho it would be pretty weird for some other objects with defaults - for example should help for @ru be the name of the contact and UUID of the flow. Ideally maybe you'd see @contac as both the current contact and an option to auto-complete as @contact as the name of the contact. Not sure.

No sure I understand.. Why would you see anything about contact in the completion for `@ru`?

**Comment by rowanseymour** [comment]
> Oh I see.. do they need to be different at all? Aren't contact and text both types fundamentally? 

Yeah those are both types... it's the difference between the part of the JSON that declares the contact type, and another part of the JSON that says this property is of type contact. I'm leaning toward the type declarations using `name` instead and then for a property to say it's of a particular type, it uses `type` which is a reference by name. That's consistent with how we use `type` elsewhere in JSON.

**Comment by rowanseymour** [comment]
> No sure I understand.. Why would you see anything about contact in the completion for @ru?

I'm saying if the help text for an object with properties should be its default, then for `@ru` it should be `name of the contact and UUID of the flow` because that's what `@run` defaults to. 

Ideally maybe we'd distinguish between the user having typed `@conta`:

<img width="282" alt="Screenshot 2019-07-10 at 12 07 30" src="https://user-images.githubusercontent.com/675558/60989382-7f773f80-a30b-11e9-9cf9-0e4220750ab7.png">

Because at that point, all we know is that they want _something_ on the contact

And the user having typed `@contact`.. showing that there are properties but also a default which in this case is name:

<img width="278" alt="Screenshot 2019-07-10 at 12 07 43" src="https://user-images.githubusercontent.com/675558/60989478-c2391780-a30b-11e9-9580-a0c8606b1f52.png">

But not sure if that's workable.

**Comment by nicpottier** [comment]
Oh I see. Ya, I think that may actually get a bit confusing honestly to
have the autocomplete change its label dynamically. (if I am understanding
correctly)

Honestly I don't care that much either way, but as I said a constant thing
we forget is how these things resolve so being explicit about it seemed
better than not to me. In a way we have been doing the generic way up to
this point (thus the confusion) which to me says maybe we should try the
explicit path for a while.

@ericnewomer any thoughts?

**Comment by rowanseymour** [comment]
I think we chatted in the past about auto-complete items having an icon to indicate whether it's got further properties to explore. That might help make it more obvious for the user who is trying to find, say contact language, that `contact` is something complex and _not just_ the name.

Could go all the way and have different tiny icons for `text|datetime|array|object`

@ericnewcomer?

**Comment by rowanseymour** [comment]
(updated the JSON so now types are declared with a `name` and properties have a `type` which references one of those types)

**Comment by ericnewcomer** [comment]
Ha, I was on the wrong repo yesterday discussing this. 

I’d definitely like the help text to be consistent regardless of how much of the match they’ve typed. I’m also okay with object icons to show the difference between leaves.

**Comment by rowanseymour** [comment]
Yeah I didn't actually mean that the help text would change. Just that for the user who has typed `@cont`, tabbing gives them `@contact` and shows the properties of contact, entering gives then `@contact` which evaluates to the name or URN of the contact and moves on. I meant that those two actions might warrant different help texts since they cater to users looking for different things.

**Comment by ericnewcomer** [comment]
I understand the distinction. Just trying to clarify what you want. So you’re saying you prefer some icon or other treatment to show that it’s not a leaf and we leave help text alone, correct?

**Comment by rowanseymour** [comment]
I think completing `@con` as just `the name or URN of the contact` or `@contact.chann` as `the name of the channel` is confusing so I'm trying to find a way to convey both 1) what the thing is and 2) what it will default to if you don't go to a leaf.

I also generally think defaults are more useful for backwards compatibility and it's actually preferable if users use more explicit leaves (e.g. `@result.foo.value` over `@result.foo`) but I think I'm on my own there.

To that end, I've currently got the completions being generated as the help text of the property + the default help text of that type if there is one.

The affected context paths are:

```
contact -> the contact (defaults to the name or URN)
contact.channel -> the preferred channel of the contact (defaults to the name)
input -> the current input from the contact (defaults to the text and attachments)
input.channel -> the channel that the input was received on (defaults to the name)
results.response_1 -> result for response_1 (defaults to the value)
run -> the current run (defaults to the contact name and flow UUID)
run.contact -> the contact of the run (defaults to the name or URN)
run.contact.channel -> the preferred channel of the contact (defaults to the name)
run.flow -> the flow of the run (defaults to the name)
child -> the last child run (defaults to the contact name and flow UUID)
child.contact -> the contact of the run (defaults to the name or URN)
child.contact.channel -> the preferred channel of the contact (defaults to the name)
child.flow -> the flow of the run (defaults to the name)
parent -> the parent of the run (defaults to the contact name and flow UUID)
parent.contact -> the contact of the run (defaults to the name or URN)
parent.contact.channel -> the preferred channel of the contact (defaults to the name)
parent.flow -> the flow of the run (defaults to the name)
```

They do seem a bit wordy but maybe we could drop all the "of the X" stuff? I dunno, brainstorming.

**Comment by ericnewcomer** [comment]
Can we put the “defaults to” bit in a different field? That way we can
tweak the treatment in the UI.

**Comment by rowanseymour** [comment]
Sure, to be clear https://github.com/nyaruka/goflow/blob/context/docs/completion.txt isn't actually intended for use by the editor - was just a way for to see a flat view of the would-be completion paths. 

So for example in `the contact (defaults to the name or URN)`..

 * `the contact` comes from the documentation of the `contact` property returned by `Run.RootContext`
 * `the name or URN` comes from the documentation of the `__default__` property returned by `Contact.Context`

**Comment by ericnewcomer** [comment]
Sorry for the confusion, looks good to me.

**Comment by rowanseymour** [comment]
Ah you meant the JSON? It already has defaults listed as properties called `__default__`

**Comment by ericnewcomer** [comment]
Yup, went back and reviewed it. Looks good.
